File: firebase/src/main/java/ro/upt/sma/heart/firebase/FirebaseHeartMeasurementDataStore.java
package ro.upt.sma.heart.firebase;

import com.google.firebase.database.ChildEventListener;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import ro.upt.sma.heart.model.HeartMeasurement;
import ro.upt.sma.heart.model.HeartMeasurementRepository;

public class FirebaseHeartMeasurementDataStore implements HeartMeasurementRepository {

  private final DatabaseReference reference;

  public FirebaseHeartMeasurementDataStore(String userId) {
    this.reference = FirebaseDatabase.getInstance().getReference().child(userId);
  }

  @Override
  public void post(HeartMeasurement heartMeasurement) {
    reference.child(heartMeasurement.timestamp + "").setValue(heartMeasurement.value);
  }

  @Override
  public void observe(final HeartChangedListener listener) {
    reference.addChildEventListener(new ChildEventListener() {
      @Override
      public void onChildAdded(DataSnapshot dataSnapshot, String s) {
        listener.onHeartChanged(toHeartMeasurement(dataSnapshot));
      }

      @Override
      public void onChildChanged(DataSnapshot dataSnapshot, String s) {
      }

      @Override
      public void onChildRemoved(DataSnapshot dataSnapshot) {
      }

      @Override
      public void onChildMoved(DataSnapshot dataSnapshot, String s) {
      }

      @Override
      public void onCancelled(DatabaseError databaseError) {
      }
    });
  }

  @Override
  public void retrieveAll(final HeartListLoadedListener listener) {
    reference.addListenerForSingleValueEvent(new ValueEventListener() {
      @Override
      public void onDataChange(DataSnapshot dataSnapshot) {
        if (dataSnapshot.hasChildren()) {
          listener.onHeartListLoaded(toHeartMeasurements(dataSnapshot.getChildren()));
        } else {
          listener.onHeartListLoaded(Collections.<HeartMeasurement>emptyList());
        }
      }

      @Override
      public void onCancelled(DatabaseError databaseError) {

      }
    });
  }


  private List<HeartMeasurement> toHeartMeasurements(Iterable<DataSnapshot> children) {
    List<HeartMeasurement> heartMeasurementList = new ArrayList<>();

    for (DataSnapshot child : children) {
      heartMeasurementList.add(toHeartMeasurement(child));
    }

    return heartMeasurementList;
  }

  private static HeartMeasurement toHeartMeasurement(DataSnapshot dataSnapshot) {
    return new HeartMeasurement(
        Long.valueOf(dataSnapshot.getKey()),
        dataSnapshot.getValue(Integer.class));
  }

}

File: data/src/main/java/ro/upt/sma/heart/firebase/FirebaseHeartMeasurementDataStore.kt
package ro.upt.sma.heart.firebase

import com.google.firebase.database.ChildEventListener
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.DatabaseReference
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.ValueEventListener
import java.util.ArrayList
import ro.upt.sma.heart.model.HeartMeasurement
import ro.upt.sma.heart.model.HeartMeasurementRepository

class FirebaseHeartMeasurementDataStore(userId: String) : HeartMeasurementRepository {

    private val reference: DatabaseReference = FirebaseDatabase.getInstance().reference.child(userId)

    override fun post(heartMeasurement: HeartMeasurement) {

        reference.child(heartMeasurement.timestamp.toString()).setValue(heartMeasurement.value)

    }

    override fun observe(listener: HeartMeasurementRepository.HeartChangedListener) {

        reference.addChildEventListener(object: ChildEventListener {
            override fun onChildAdded(snapshot: DataSnapshot, string: String?) {
                listener.onHeartChanged(toHeartMeasurement(snapshot))
            }

            override fun onChildChanged(snapshot: DataSnapshot, string: String?) { }

            override fun onChildRemoved(snapshot: DataSnapshot) { }

            override fun onChildMoved(snapshot: DataSnapshot, string: String?) { }

            override fun onCancelled(snapshot: DatabaseError) { }

        })

    }

    override fun retrieveAll(listener: HeartMeasurementRepository.HeartListLoadedListener) {
        reference.addListenerForSingleValueEvent(object: ValueEventListener {
            override fun onDataChange(snapshot: DataSnapshot) {
                val children = snapshot.children

                listener.onHeartListLoaded(toHeartMeasurements(children))

                reference.removeEventListener(this)
            }

            override fun onCancelled(snapshot: DatabaseError) {
                reference.removeEventListener(this)
            }
        })
    }


    private fun toHeartMeasurements(children: Iterable<DataSnapshot>): List<HeartMeasurement> {
        val heartMeasurementList = ArrayList<HeartMeasurement>()

        for (child in children) {
            heartMeasurementList.add(toHeartMeasurement(child))
        }

        return heartMeasurementList
    }

    private fun toHeartMeasurement(dataSnapshot: DataSnapshot): HeartMeasurement {
        return HeartMeasurement(
                java.lang.Long.valueOf(dataSnapshot.key),
                dataSnapshot.getValue(Int::class.java)!!)
    }

}
